wl/server/TransifexIntegration.java

```diff
@@ -0,0 +1,256 @@
+import org.json.simple.parser.*;
+import wl.utils.*;
+
+public class TransifexIntegration {
+	public static TransifexIntegration TX = new TransifexIntegration();
+	private TransifexIntegration() {}
+
+	public synchronized void fullSync() throws Exception {
+		pull();
+		buildCatalogues();
+		push();
+	}
+
+	public synchronized void pull() throws Exception {
+		ServerUtils.log("Pulling translations from Transifex...");
+		Utils.bashOutput("tx", "pull", "-f", "-a");
+	}
+	public synchronized void push() throws Exception {
+		ServerUtils.log("Pushing POT files to Transifex...");
+		Utils.bashOutput("tx", "push", "-s");
+	}
+
+	public synchronized void buildCatalogues() throws Exception {
+		ServerUtils.log("Rebuilding catalogues...");
+		Buildcats.buildCatalogues();
+
+		ServerUtils.log("Updating PO and MO files...");
+		Utils.bashOutput("rm", "-r", "i18n");
+		for (File poDir : Utils.listSorted(new File("po"))) {
+			for (File poFile : Utils.listSorted(poDir)) {
+				if (poFile.getName().endsWith(".po")) {
+					final String lang =
+					    poFile.getName().substring(0, poFile.getName().length() - 3);
+					final String outFile = "i18n/" + poDir.getName() + "/" + lang + ".mo";
+
+					Utils.bashOutput("msgmerge", poFile.getAbsolutePath(),
+					                 new File(poDir, poDir.getName() + ".pot").getAbsolutePath(),
+					                 "-o", poFile.getAbsolutePath());
+					Utils.bashOutput("mkdir", "-p", "i18n/" + poDir.getName());
+					Utils.bashOutput("mkdir", "-p", "i18n/" + lang + "/LC_MESSAGES");
+					Utils.bashOutput("msgfmt", poFile.getAbsolutePath(), "-o", outFile);
+					// We permanently need to store MO files in two different
+					// locations for backwards compatibility.
+					Utils.bashOutput(
+					    "cp", outFile, "i18n/" + lang + "/LC_MESSAGES/" + poDir.getName() + ".mo");
+				}
+			}
+		}
+
+		ServerUtils.log("Gathering translation changes...");
+		List<String> changedMO = new ArrayList<>();
+		Utils.bashOutput("git", "add", "i18n");
+		for (String changed :
+		     Utils.bashOutput("bash", "-c", "git status -s i18n/*.wad").split("\n")) {
+			if (changed.trim().isEmpty()) continue;
+			String[] split = changed.split(" ");  // "", "M", "i18n/fishy.wad/nds.mo"
+			changed = split[split.length - 1];    // "i18n/fishy.wad/nds.mo"
+			split = changed.split("/");           // "i18n", "fishy.wad", "nds.mo"
+			changedMO.add("+" + split[1]);        // "+fishy.wad"
+		}
+		UpdateList.main(changedMO.toArray(new String[0]));
+		Utils._staticprofiles.clear();
+		Utils.bashOutput("./skip_timestamp_only_po_changes.sh");
+	}
+
+	private static class Issue {
+		public final String issueID, message, string, stringID, occurrence, addon;
+
+		private Issue(
+		    String issueID, String message, String string, String stringID, String occurrence) {
+			this.issueID = issueID;
+			this.message = message;
+			this.string = string;
+			this.stringID = stringID;
+			this.occurrence = occurrence;
+
+			addon = occurrence.split("/")[1];
```

Comment:
Is this safe?

Reply:
It should always be unless the API call returned garbage (or Transifex changes the format). In this very bad case the IndexOutOfBoundsException throws us all the way out up to the `SyncThread.run` catch clause, which sends a notification that something has gone very wrong. This is desired because Tx integration issues would be hard to detect if this didn't complain very verbosely.